Port gdb-server to use the gdbstub crate

thefaxman opened the pull request:
* Replace existing GDB parsing logic with gdbstub
* Add support for multi-core debugging via the multi-threaded gdbstub interfaces
* Fix session setup to run DebugCoreStart on each core instead of just the first one

Closes #608

Signed-off-by: Ryan Fairfax <ryan@thefaxman.net>

Yatekii reviewed (changes requested):
Looking very good, thanks! :)

I added some minor nits here and there and some questions :)

Yatekii commented:
Btw, @thefaxman, did you see https://github.com/probe-rs/probe-rs/issues/608?

thefaxman commented on gdb-server/src/arch.rs:
```diff
@@ -0,0 +1,61 @@
+
+/// Generic implementation for a runtime evaluated architecture
+impl Arch for RuntimeArch {
+    // We can always handle a 64-bit address space
```
This is just for the internal types used by the stub.  GDB knows the architecture so it'll bound any input / output to the right size for the pc register so it will never ask for something out of bounds.

thefaxman commented on gdb-server/src/stub.rs:
```diff
@@ -0,0 +1,125 @@
+/// Configuration for a single GDB endpoint
+pub struct GdbInstanceConfiguration {
+    /// The core type that will be sent to GDB
+    pub core_type: CoreType,
```
Correct.  The stub sends a single architecture that all processes / threads must match.  OpenOCD and J-Link do what I'm doing here and stand up multiple endpoints for each architecture.

Modern GDB clients can connect to multiple stub endpoints to make this a bit better, but it's still somewhat rough.

thefaxman commented on gdb-server/src/target/breakpoints.rs:
```diff
@@ -0,0 +1,53 @@
+    ) -> gdbstub::target::TargetResult<bool, Self> {
+        let mut session = self.session.borrow_mut();
+
+        for core_id in &self.cores {
```
GDB expects that breakpoints apply to all threads - aka cores in our world.

thefaxman commented on gdb-server/src/target/monitor.rs:
```diff
@@ -0,0 +1,30 @@
+            _ => {
+                outputln!(out, "{}", HELP_TEXT);
+            }
+        }
```
Yes, there's a trait for that path that should be much easier to add next.  I may have some cycles to look at this after fixing up some FP items.

thefaxman commented on gdb-server/src/target/breakpoints.rs:
```diff
@@ -0,0 +1,53 @@
+impl HwBreakpoint for RuntimeTarget<'_> {
+    fn add_hw_breakpoint(
+        &mut self,
+        addr: <Self::Arch as gdbstub::arch::Arch>::Usize,
```
I was surprised as well, but without this it trips Rust error E0223.  I think this is a limit in the compiler's deduction logic.

https://doc.rust-lang.org/error-index.html#E0223

daniel5151 replied:
quick drive-by comment: you can swap out this long name for the concrete type being used (in this case, `u64`). See https://docs.rs/gdbstub/latest/gdbstub/target/index.html#a-note-on-all-the-selfarch-as-arch-syntax

thefaxman replied:
Much cleaner, thanks.  I've updated all the `Arch::<X>` references to the final types.

thefaxman commented on gdb-server/src/stub.rs:
```diff
@@ -0,0 +1,125 @@
+
+const CONNECTION_STRING: &str = "127.0.0.1:1337";
+
+fn adjust_addrs(addrs: &[SocketAddr], offset: usize) -> Vec<SocketAddr> {
```
Sure thing, I'll add a number of comments to try to clear this up and move the helper to the bottom.

thefaxman commented on gdb-server/src/target/desc.rs:
```diff
@@ -0,0 +1,65 @@
+        buf: &mut [u8],
+    ) -> gdbstub::target::TargetResult<usize, Self> {
+        let mut session = self.session.borrow_mut();
+        let xml = session.gdb_memory_map(self.cores[0]).into_target_result()?;
```
Yes, moved the logic out of the extension traits and into these files since it's now only used from one exactly one module.

thefaxman commented:
> Btw, @thefaxman, did you see #608?

Yes, that's what led me to investigate gdbstub as an option.  I updated the commit message to tag that it closes that issue.

I believe I've addressed all the feedback - thanks for the review.

Yatekii reviewed (commented):
Thanks a lot for all the hard work put into this :)

Hmm did we test this on 32bit MCUs, just to be sure? :/

Yatekii reviewed (approved):
bors r+

bugadani commented on gdb-server/src/stub.rs:
```diff
@@ -0,0 +1,132 @@
+        let mut wait_time = Duration::ZERO;
+
+        for target in targets.iter_mut() {
+            wait_time = wait_time.min(target.process()?);
```
Can this `min` ever produce something other than `Duration::ZERO`?

Yatekii replied:
Maybe it should have been a max? Not sure. I would agree that the min is useless if `process()` returns strictly positive numbers.
